File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    TreeNode *tree(vector<int> &preorder, int pstart, int pend, vector<int> &inorder, int instart, int inend, map<int,int> &mpp){
        if (instart > inend || pstart > pend){
            return nullptr;
        }

        TreeNode *root = new TreeNode(preorder[pstart]);
        int inroot = mpp[root->val];
        int nleft = inroot - instart;

        root->left = tree(preorder, pstart+1, pstart+nleft, inorder, instart, inroot-1, mpp);
        root->right = tree(preorder, pstart+nleft+1, pend, inorder, inroot+1, inend, mpp);

        return root;
    }
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        map<int,int> mpp;
        for (int i=0; i<inorder.size(); i+=1){
            mpp[inorder[i]] = i;
        }

        TreeNode *root = tree(preorder, 0, preorder.size()-1, inorder, 0, inorder.size()-1, mpp);
        return root;
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func builder(preorder []int, pst, pend int, inorder []int, inst, inend int, idxmp map[int]int) *TreeNode {
    if pst > pend || inst > inend {
        return nil
    }

    root := &TreeNode{Val: preorder[pst]}
    inroot := idxmp[root.Val]
    left := inroot - inst

    root.Left = builder(preorder, pst+1, pst+left, inorder, inst, inroot-1,idxmp)
    root.Right = builder(preorder, pst+left+1, pend, inorder, inroot+1, inend, idxmp)

    return root
}

func buildTree(preorder []int, inorder []int) *TreeNode {
    idxmp := make(map[int]int)

    for i := 0; i < len(inorder); i++ {
        idxmp[inorder[i]] = i
    }

    return builder(preorder, 0, len(preorder)-1, inorder, 0, len(inorder)-1, idxmp)
}